### ISSUE TITLE
False positive on self within callable used in CallbackPromise (prophecy)

### ISSUE THREAD

**Issue by iltar** [kalessil/phpinspectionsea#564]
The following code causes an issue:
![image](https://user-images.githubusercontent.com/1754678/31119653-f1245568-a831-11e7-9ad2-cc4521fbb020.png)

If I replace this with self, I get this error:
```
1) MyTest::testSomething
Error: Call to undefined method Prophecy\Prophecy\ObjectProphecy::assertThat()
```

The problem is that `assertInstanceOf` is defined in the parent (as static) and then calls the `assertThat`, which is static and is bound to the ObjectProphecy instead of the test due to how prophecy binds the callable:

```php
    public function execute(array $args, ObjectProphecy $object, MethodProphecy $method)
    {
        $callback = $this->callback;

        if ($callback instanceof Closure && method_exists('Closure', 'bind')) {
            $callback = Closure::bind($callback, $object);
        }

        return call_user_func($callback, $args, $object, $method);
    }
```

So in this particular scenario, because of the scope binding, this check returns a false positive.

**Comment by kalessil** [comment]
Can this be related to #557 and #555?

**Comment by iltar** [comment]
Doesn't look like it, but I can't be sure. The reason this fails, is because of the change of the scope of the closure, which binds it to the prophecy instead of the test. This is behavior you cannot predict and you can't resolve easily as it can go through many layers of abstraction.

Personally I would simply turn this particular check off if it's in a closer that's not directly used within the same scope.

**Comment by kalessil** [comment]
Huh, we need to skip closures analysis then, thank you for reporting @iltar

**Comment by iltar** [comment]
It works in most scenarios, just not in this one. I'm considering opening a bug report at phpunit as well, because imo it could be classified as a bug in phpunit. 

Code to reproduce this:
```php
// phpunit/phpunit 6.3.0
use PHPUnit\Framework\TestCase;

interface TestInterface
{
    public function stub();
}

class SimpleTest extends TestCase
{
    public function test()
    {
        $prophecy = $this->prophesize(TestInterface::class);
        $prophecy->stub()->will(function() {
            var_dump(static::class);
            var_dump(self::class);
        });

        $prophecy->reveal()->stub();
    }
}

// outputs
// string(32) "Prophecy\Prophecy\ObjectProphecy"
// string(22) "SimpleTest"
```

So calling "self" works 100% fine, but if the "self" called methods call other methods in a "static" context, it breaks. I have been unable to reproduce this writing code myself, so it seems more of an edge-case :thinking:

**Comment by kalessil** [comment]
Okay, then please ping me back when the related PhpUnit ticket reaches any sort of agreement if this bug or feature =)

**Comment by iltar** [comment]
So it's not a bug in phpunit, but it _is_ something that's really hard to detect using static code analysis. My suggestion would be to skip closures _unless_ you can guarantee that the scope is not changed using `\Closure::bind` or `\Closure::bindTo`

**Comment by kalessil** [comment]
Ok, then we'll just skip closure-scope for the inspection.
